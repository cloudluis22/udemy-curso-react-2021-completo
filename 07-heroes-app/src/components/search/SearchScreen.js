import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    console.log(q);

    const [formValues, handleInputChange, reset] = useForm({
        searchText: q
    });

    const { searchText } = formValues;

   const heroesFiltered = useMemo(() => getHeroesByName( q ), [q]);

    const handleSearch = (evt) => {
        evt.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return (
        <div>

            <h1> Search Screen </h1>
            <hr />

            <div className="row">

                <div className="col-4">

                    <h4> Search Form </h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Search Hero"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <div className="d-grid gap-2 mt-2">
                            <button type="submit" className="btn btn-outline-primary"> Search </button>
                        </div>

                    </form>

                </div>

                <div className="col-7">
                    <h4> Results </h4>
                    <hr />

                    {
                        (q === '') &&
                        <div className="alert alert-info"> Search a hero </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                        <div className="alert alert-danger"> { q } not found, try another search </div>
                    }

                    {

                        heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }

                </div>
            </div>

        </div>
    )
}