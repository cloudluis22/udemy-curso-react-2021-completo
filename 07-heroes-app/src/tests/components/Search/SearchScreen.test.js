import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en el componente de "<SearchScreen/>".', () => {

    test('Debe de mostrarse con los valores por defecto.', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');

    });

    test('Debe de mostrar a Batman y el input con el valor del query string.', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');

    });

    test('Debe de mostrar un error si no se encuentra el hero.', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').exists()).toBe(true);

    });

    test('should ', () => {

        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']} >
                <Route path="/search" component={() => <SearchScreen history={ history } />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault() { }
        });

        expect(history.push).toHaveBeenCalledWith('?q=batman');

    });

});