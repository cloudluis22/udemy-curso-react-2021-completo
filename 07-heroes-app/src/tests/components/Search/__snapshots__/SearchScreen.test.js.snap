// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente de "<SearchScreen/>". Debe de mostrarse con los valores por defecto. 1`] = `
<div>
  <h1>
     Search Screen 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2 mt-2"
        >
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
             Search 
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
         Search a hero 
      </div>
    </div>
  </div>
</div>
`;
